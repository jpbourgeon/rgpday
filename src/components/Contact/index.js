import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import { ReCaptcha } from 'recaptcha-v3-react'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import Button from '@material-ui/core/Button'
import Snackbar from '@material-ui/core/Snackbar'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import Email from '@material-ui/icons/EmailOutlined'
import isEmail from 'validator/lib/isEmail'
import isEmpty from 'validator/lib/isEmpty'
import AppBar from '../AppBar'

const styles = theme => {
  return {
    '@global': {
      '.grecaptcha-badge': {
        visibility: 'visible !important'
      }
    },
    layout: {
      width: 'auto',
      marginLeft: theme.spacing.unit * 3,
      marginRight: theme.spacing.unit * 3,
      [theme.breakpoints.up(1100 + theme.spacing.unit * 3 * 2)]: {
        width: 1100,
        marginLeft: 'auto',
        marginRight: 'auto'
      }
    },
    titleIcon: {
      fontSize: '2.5rem',
      marginBottom: '-0.5rem'
    },
    formContainer: {
      display: 'flex',
      flexWrap: 'wrap'
    },
    mainFeaturedPost: {
      backgroundSize: `100% 100%`,
      marginBottom: theme.spacing.unit * 4
    },
    mainFeaturedPostContent: {
      padding: `${theme.spacing.unit * 6}px`
    },
    button: {
      margin: theme.spacing.unit
    },
    closeSnackbar: {
      padding: theme.spacing.unit / 2
    }
  }
}

const defaultState = {
  content: {
    isDirty: false,
    value: ''
  },
  from: {
    isDirty: false,
    value: ''
  },
  form: {
    isDisabled: false
  },
  recaptcha: {
    token: ''
  },
  snackbar: {
    open: false,
    message: ''
  },
  subject: {
    isDirty: false,
    value: ''
  }
}

class Contact extends React.Component {
  constructor (props) {
    super(props)
    this.state = { ...defaultState }

    this.handleChange = this.handleChange.bind(this)
    this.handleBlur = this.handleBlur.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleCancel = this.handleCancel.bind(this)
    this.handleCloseSnackbar = this.handleCloseSnackbar.bind(this)
    this.handleReCaptchaToken = this.handleReCaptchaToken.bind(this)
  }

  handleChange (event, field) {
    const state = {}
    state[field] = {
      isDirty: this.state[field].isDirty,
      value: event.target.value
    }
    this.setState({ ...state })
  }

  handleBlur (event, field) {
    const state = {}
    state[field] = {
      isDirty: true,
      value: this.state[field].value
    }
    this.setState({ ...state })
  }

  handleSubmit (event) {
    event.preventDefault()
    const snackbar = {
      open: true,
      message: ''
    }
    if (
      isEmail(this.state.from.value) &&
      !isEmpty(this.state.from.value) &&
      !isEmpty(this.state.subject.value) &&
      !isEmpty(this.state.content.value) &&
      !isEmpty(this.state.recaptcha.token)
    ) {
      // TODO : Call SES Lambda and manage response (success if recaptcha verified and email sent, fail otherwise)
      // Disable form during Lambda call
      // Reset form on success but not on failure !
      snackbar.message = `Votre message a été envoyé avec succès.`
      this.setState({ snackbar })
    } else {
      snackbar.message = `Votre message n'a pas été envoyé. Le formulaire est invalide.`
      this.setState({ snackbar })
    }
  }

  handleCancel (event) {
    event.preventDefault()
    const state = { ...defaultState, recaptcha: this.state.recaptcha }
    this.setState({ ...state })
  }

  handleCloseSnackbar (event, reason) {
    if (reason === 'clickaway') {
      return
    }
    const snackbar = {
      open: false,
      message: ''
    }
    this.setState({ snackbar })
  }

  handleReCaptchaToken (token) {
    const recaptcha = { token }
    this.setState({ recaptcha })
  }

  render () {
    const { classes } = this.props
    return (
      <div className={classes.layout}>
        <AppBar />
        <main>
          <Paper className={classes.mainFeaturedPost}>
            <Grid container>
              <Grid item xs={12}>
                <div className={classes.mainFeaturedPostContent}>
                  <Typography variant='h4' color='inherit' gutterBottom>
                    <Email className={classes.titleIcon} />&nbsp;Contact
                  </Typography>

                  <form
                    className={classes.formContainer}
                    noValidate
                    autoComplete='off'
                    onSubmit={this.handleSubmit}
                  >
                    <TextField
                      label='Votre email'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'from')}
                      onBlur={(e) => this.handleBlur(e, 'from')}
                      value={this.state.from.value}
                      error={(
                        this.state.from.isDirty &&
                        (
                          !isEmail(this.state.from.value) ||
                          isEmpty(this.state.from.value)
                        )
                      )}
                      disabled={this.state.form.isDisabled}
                    />
                    <TextField
                      label='Sujet'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'subject')}
                      onBlur={(e) => this.handleBlur(e, 'subject')}
                      value={this.state.subject.value}
                      error={(this.state.subject.isDirty && isEmpty(this.state.subject.value))}
                      disabled={this.state.form.isDisabled}
                    />
                    <TextField
                      label='Description'
                      multiline
                      rows='10'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'content')}
                      onBlur={(e) => this.handleBlur(e, 'content')}
                      value={this.state.content.value}
                      error={(this.state.content.isDirty && isEmpty(this.state.content.value))}
                      disabled={this.state.form.isDisabled}
                    />
                    <Button
                      variant='outlined'
                      color='secondary'
                      className={classes.button}
                      type='submit'
                      disabled={this.state.form.isDisabled}
                    >
                    Envoyer
                    </Button>
                    <Button
                      variant='outlined'
                      className={classes.button}
                      onClick={this.handleCancel}
                      disabled={this.state.form.isDisabled}
                    >
                    Annuler
                    </Button>
                  </form>

                  <Snackbar
                    anchorOrigin={{
                      vertical: 'bottom',
                      horizontal: 'center'
                    }}
                    open={this.state.snackbar.open}
                    autoHideDuration={5000}
                    onClose={this.handleClose}
                    ContentProps={{
                      'aria-describedby': 'message-id'
                    }}
                    message={<span id='message-id'>{this.state.snackbar.message}</span>}
                    action={[
                      <IconButton
                        key='close'
                        aria-label='Close'
                        color='inherit'
                        className={classes.closeSnackbar}
                        onClick={this.handleCloseSnackbar}
                      >
                        <CloseIcon />
                      </IconButton>
                    ]}
                  />
                </div>
              </Grid>
            </Grid>
          </Paper>
        </main>
        <ReCaptcha
          action='contact'
          sitekey='6LedLpMUAAAAAG8Ai4M4x9wTcIs4rPmvYV82a7Yh'
          verifyCallback={this.handleReCaptchaToken}
        />
      </div>
    )
  }
}

Contact.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Contact)
