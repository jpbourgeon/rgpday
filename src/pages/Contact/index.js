import React from 'react'
import logger from 'src/logger'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import ReCaptcha from 'src/components/ReCaptcha'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import Button from '@material-ui/core/Button'
import Snackbar from '@material-ui/core/Snackbar'
import IconButton from '@material-ui/core/IconButton'
import CloseIcon from '@material-ui/icons/Close'
import Email from '@material-ui/icons/EmailOutlined'
import isEmail from 'validator/lib/isEmail'
import isEmpty from 'validator/lib/isEmpty'

import API, { graphqlOperation } from '@aws-amplify/api'
import { sendMail } from 'src/graphql/queries'
import config from 'src/aws-exports'

API.configure(config)

const styles = theme => {
  return {
    '@global': {
      '.grecaptcha-badge': {
        visibility: 'visible'
      }
    },
    layout: {
      width: 'auto',
      marginLeft: theme.spacing.unit * 3,
      marginRight: theme.spacing.unit * 3,
      [theme.breakpoints.up(1100 + theme.spacing.unit * 3 * 2)]: {
        width: 1100,
        marginLeft: 'auto',
        marginRight: 'auto'
      }
    },
    titleIcon: {
      fontSize: '2.5rem',
      marginBottom: '-0.5rem'
    },
    formContainer: {
      display: 'flex',
      flexWrap: 'wrap'
    },
    mainFeaturedPost: {
      backgroundSize: `100% 100%`,
      marginBottom: theme.spacing.unit * 4
    },
    mainFeaturedPostContent: {
      padding: `${theme.spacing.unit * 6}px`
    },
    button: {
      marginRight: theme.spacing.unit,
      marginTop: theme.spacing.unit * 2
    },
    closeSnackbar: {
      padding: theme.spacing.unit / 2
    }
  }
}

const defaultState = {
  content: {
    isDirty: false,
    value: ''
  },
  sender: {
    isDirty: false,
    value: ''
  },
  form: {
    isDisabled: false
  },
  snackbar: {
    open: false,
    message: ''
  },
  subject: {
    isDirty: false,
    value: ''
  }
}

class Contact extends React.Component {
  constructor (props) {
    super(props)
    this._isMounted = false
    this.state = { ...defaultState }
    this.ReCaptchaRef = React.createRef()

    this.handleChange = this.handleChange.bind(this)
    this.handleBlur = this.handleBlur.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleCancel = this.handleCancel.bind(this)
    this.handleCloseSnackbar = this.handleCloseSnackbar.bind(this)
    this.handleReCaptchaToken = this.handleReCaptchaToken.bind(this)
  }

  componentDidMount () {
    this._isMounted = true
  }

  componentWillUnmount () {
    this._isMounted = false
  }

  handleChange (event, field) {
    const state = {}
    state[field] = {
      isDirty: this.state[field].isDirty,
      value: event.target.value
    }
    this.setState({ ...state })
  }

  handleBlur (event, field) {
    const state = {}
    state[field] = {
      isDirty: true,
      value: this.state[field].value
    }
    this.setState({ ...state })
  }

  async handleSubmit (event) {
    event.preventDefault()
    let state = this.state
    if (
      !isEmpty(this.state.sender.value) && isEmail(this.state.sender.value) &&
      !isEmpty(this.state.subject.value) &&
      !isEmpty(this.state.content.value)
    ) {
      state.form.isDisabled = true
      state.snackbar.message = `Votre message est en cours d'envoi. Merci de patienter...`
      state.snackbar.open = true
      if (this._isMounted) {
        this.setState({ state }, () => {
          this.ReCaptchaRef.current.execute()
        })
      }
    } else {
      state.snackbar.message = `Votre message n'a pas été envoyé. Le formulaire est invalide.`
      state.snackbar.open = true
      if (this._isMounted) this.setState({ state })
    }
  }

  handleCancel (event = { preventDefault: () => {} }) {
    event.preventDefault()
    this.setState({ ...defaultState, form: { isDisabled: false }, snackbar: { open: false, message: '' } })
  }

  handleCloseSnackbar (event, reason) {
    if (reason === 'clickaway') {
      return
    }
    const snackbar = {
      open: false,
      message: ''
    }
    this.setState({ snackbar })
  }

  async handleReCaptchaToken (token) {
    let state = this.state
    try {
      const result = await API.graphql(
        graphqlOperation(sendMail, {
          'sender': this.state.sender.value,
          'content': this.state.content.value,
          'subject': this.state.subject.value,
          'recaptcha': token
        })
      )
      logger.info(result)
      if (!result.errors) {
        logger.info('handleSubmit', result)
        state = { ...defaultState }
        state.snackbar.message = `Votre message a été envoyé avec succès.`
        state.snackbar.open = true
      } else {
        logger.error('handleSubmit', result)
        state.snackbar.message = `L'envoi de votre message a échoué.`
        state.snackbar.open = true
      }
    } catch (error) {
      logger.error('handleSubmit', error)
      state.snackbar.message = `L'envoi de votre message a échoué.`
      state.snackbar.open = true
    } finally {
      if (this._isMounted) this.setState({ ...state, form: { isDisabled: false } })
    }
  }

  render () {
    const { classes } = this.props
    return (
      <div className={classes.layout}>
        <main>
          <Paper className={classes.mainFeaturedPost}>
            <Grid container>
              <Grid item xs={12}>
                <div className={classes.mainFeaturedPostContent}>
                  <Typography variant='h4' color='inherit' gutterBottom>
                    <Email className={classes.titleIcon} />&nbsp;Contact
                  </Typography>

                  <form
                    className={classes.formContainer}
                    noValidate
                    autoComplete='off'
                    onSubmit={this.handleSubmit}
                  >
                    <TextField
                      label='Votre email *'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'sender')}
                      onBlur={(e) => this.handleBlur(e, 'sender')}
                      value={this.state.sender.value}
                      error={(
                        this.state.sender.isDirty &&
                        (
                          !isEmail(this.state.sender.value) ||
                          isEmpty(this.state.sender.value)
                        )
                      )}
                      disabled={this.state.form.isDisabled}
                    />
                    <TextField
                      label='Sujet *'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'subject')}
                      onBlur={(e) => this.handleBlur(e, 'subject')}
                      value={this.state.subject.value}
                      error={(this.state.subject.isDirty && isEmpty(this.state.subject.value))}
                      disabled={this.state.form.isDisabled}
                    />
                    <TextField
                      label='Message *'
                      multiline
                      rows='10'
                      fullWidth
                      margin='normal'
                      variant='outlined'
                      onChange={(e) => this.handleChange(e, 'content')}
                      onBlur={(e) => this.handleBlur(e, 'content')}
                      value={this.state.content.value}
                      error={(this.state.content.isDirty && isEmpty(this.state.content.value))}
                      disabled={this.state.form.isDisabled}
                    />
                    <Button
                      variant='outlined'
                      color='secondary'
                      className={classes.button}
                      type='submit'
                      disabled={this.state.form.isDisabled}
                    >
                    Envoyer
                    </Button>
                    <Button
                      variant='outlined'
                      className={classes.button}
                      onClick={this.handleCancel}
                      disabled={this.state.form.isDisabled}
                    >
                    Annuler
                    </Button>
                  </form>

                  <Snackbar
                    anchorOrigin={{
                      vertical: 'bottom',
                      horizontal: 'center'
                    }}
                    open={this.state.snackbar.open}
                    autoHideDuration={10000}
                    onClose={this.handleCloseSnackbar}
                    ContentProps={{
                      'aria-describedby': 'message-id'
                    }}
                    message={<span id='message-id'>{this.state.snackbar.message}</span>}
                    action={[
                      <IconButton
                        key='close'
                        aria-label='Close'
                        color='inherit'
                        className={classes.closeSnackbar}
                        onClick={this.handleCloseSnackbar}
                      >
                        <CloseIcon />
                      </IconButton>
                    ]}
                  />
                </div>
              </Grid>
            </Grid>
          </Paper>
        </main>
        <ReCaptcha
          action='contact'
          sitekey='6LedLpMUAAAAAG8Ai4M4x9wTcIs4rPmvYV82a7Yh'
          verifyCallback={this.handleReCaptchaToken}
          ref={this.ReCaptchaRef}
        />
      </div>
    )
  }
}

Contact.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Contact)
