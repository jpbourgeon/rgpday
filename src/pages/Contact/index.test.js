/* global describe, it, expect */
import React from 'react'
import renderer from 'react-test-renderer'
import Contact from './'
import jssSerializer from 'src/utils/jssSnapshotSerializer'

expect.addSnapshotSerializer(jssSerializer)

describe('The Contact page', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Contact />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
