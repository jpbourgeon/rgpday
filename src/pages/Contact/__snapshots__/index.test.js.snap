// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Contact page should render correctly 1`] = `
<div
  className="layout-1"
>
  <main>
    <div
      className="root-8 elevation2-12 rounded-9 mainFeaturedPost-4"
    >
      <div
        className="container-35"
      >
        <div
          className="item-36 12-75"
        >
          <div
            className="mainFeaturedPostContent-5"
          >
            <h4
              className="root-132 h4-147 colorInherit-161 gutterBottom-159"
            >
              <svg
                aria-hidden="true"
                className="root-168 titleIcon-2"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0V0z"
                  fill="none"
                />
                <g>
                  <path
                    d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0l-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
                  />
                </g>
              </svg>
              Â Contact
            </h4>
            <form
              autoComplete="off"
              className="formContainer-3"
              noValidate={true}
              onSubmit={[Function]}
            >
              <div
                className="root-177 marginNormal-178 fullWidth-180"
              >
                <label
                  className="root-192 root-181 formControl-186 animated-189 outlined-191"
                  data-shrink={false}
                >
                  Votre email *
                </label>
                <div
                  className="root-212 root-199 fullWidth-221 formControl-213"
                  onClick={[Function]}
                >
                  <fieldset
                    aria-hidden={true}
                    className="root-229 notchedOutline-206"
                    style={
                      Object {
                        "paddingLeft": 8,
                      }
                    }
                  >
                    <legend
                      className="legend-230"
                      style={
                        Object {
                          "width": 0.01,
                        }
                      }
                    >
                      <span
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                  <input
                    aria-invalid={false}
                    className="input-222 input-207"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="root-177 marginNormal-178 fullWidth-180"
              >
                <label
                  className="root-192 root-181 formControl-186 animated-189 outlined-191"
                  data-shrink={false}
                >
                  Sujet *
                </label>
                <div
                  className="root-212 root-199 fullWidth-221 formControl-213"
                  onClick={[Function]}
                >
                  <fieldset
                    aria-hidden={true}
                    className="root-229 notchedOutline-206"
                    style={
                      Object {
                        "paddingLeft": 8,
                      }
                    }
                  >
                    <legend
                      className="legend-230"
                      style={
                        Object {
                          "width": 0.01,
                        }
                      }
                    >
                      <span
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                  <input
                    aria-invalid={false}
                    className="input-222 input-207"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="root-177 marginNormal-178 fullWidth-180"
              >
                <label
                  className="root-192 root-181 formControl-186 animated-189 outlined-191"
                  data-shrink={false}
                >
                  Message *
                </label>
                <div
                  className="root-212 root-199 fullWidth-221 formControl-213 multiline-220 multiline-205"
                  onClick={[Function]}
                >
                  <fieldset
                    aria-hidden={true}
                    className="root-229 notchedOutline-206"
                    style={
                      Object {
                        "paddingLeft": 8,
                      }
                    }
                  >
                    <legend
                      className="legend-230"
                      style={
                        Object {
                          "width": 0.01,
                        }
                      }
                    >
                      <span
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                  <textarea
                    aria-invalid={false}
                    className="input-222 input-207 inputMultiline-224 inputMultiline-209"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    rows="10"
                    value=""
                  />
                </div>
              </div>
              <button
                className="root-257 root-231 outlined-239 outlinedSecondary-241 button-6"
                disabled={false}
                onBlur={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="submit"
              >
                <span
                  className="label-232"
                >
                  Envoyer
                </span>
                <span
                  className="root-267"
                />
              </button>
              <button
                className="root-257 root-231 outlined-239 button-6"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="label-232"
                >
                  Annuler
                </span>
                <span
                  className="root-267"
                />
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
